source/ncengine/graphics2/diligent/pass/WireframePass.cpp

```diff
@@ -0,0 +1,172 @@
+             Diligent::IPipelineResourceSignature& perFrameResourceSignature)
+{
+    using namespace Diligent;
+
+    auto vertexShader = shaderFactory.MakeShaderFromSource(
+        std::span{g_vertexShader},
+        "Wireframe VS",
+        Diligent::SHADER_TYPE_VERTEX,
+        Diligent::SHADER_SOURCE_LANGUAGE_HLSL
+    );
+
+    auto pixelShader = shaderFactory.MakeShaderFromSource(
+        std::span{g_pixelShader},
+        "Wireframe PS",
+        Diligent::SHADER_TYPE_PIXEL,
+        Diligent::SHADER_SOURCE_LANGUAGE_HLSL
+    );
+
+    auto signatures = std::array{&perFrameResourceSignature};
+    auto layoutElements = nc::graphics::GetMeshVertexLayoutElements(0);
+
+    auto ci = GraphicsPipelineStateCreateInfo{};
+    ci.PSODesc.PipelineType = PIPELINE_TYPE_GRAPHICS;
+    ci.PSODesc.Name = "WireframePipeline";
+    ci.PSODesc.ResourceLayout.DefaultVariableType = SHADER_RESOURCE_VARIABLE_TYPE_STATIC;
+
+    ci.ppResourceSignatures = signatures.data();
+    ci.ResourceSignaturesCount = static_cast<uint32_t>(signatures.size());
+
+    ci.pVS = vertexShader;
+    ci.pPS = pixelShader;
+
+    ci.GraphicsPipeline.NumRenderTargets             = 1;
+    ci.GraphicsPipeline.RTVFormats[0]                = swapChain.GetDesc().ColorBufferFormat;
+    ci.GraphicsPipeline.DSVFormat                    = swapChain.GetDesc().DepthBufferFormat;
+    ci.GraphicsPipeline.PrimitiveTopology            = PRIMITIVE_TOPOLOGY_TRIANGLE_LIST;
+    ci.GraphicsPipeline.RasterizerDesc.CullMode      = CULL_MODE_BACK;
+    ci.GraphicsPipeline.DepthStencilDesc.DepthEnable = True;
+    ci.GraphicsPipeline.InputLayout.LayoutElements   = layoutElements.data();
+    ci.GraphicsPipeline.InputLayout.NumElements      = static_cast<uint32_t>(layoutElements.size());
+    ci.GraphicsPipeline.RasterizerDesc.FillMode      = FILL_MODE_WIREFRAME;
+
+    auto pso = RefCntAutoPtr<IPipelineState>{};
+    device.CreateGraphicsPipelineState(ci, &pso);
+    if (!pso)
+    {
+        throw nc::NcError{"Failed to create wireframe pipeline"};
+    }
+
+    return pso;
+}
+} // anonymous namespace
+
+namespace nc::graphics
+{
+WireframePass::WireframePass(Diligent::IRenderDevice& device,
+                             Diligent::ISwapChain& swapChain,
+                             ShaderFactory& shaderFactory,
+                             ShaderBindings& shaderBindings)
+    : m_pso{MakePso(device, swapChain, shaderFactory, shaderBindings.GetPerFrameSignature().GetResourceSignature())},
+      m_buffer{&shaderBindings.GetPerFrameSignature().GetWireframeBuffer()}
+{
+}
+
+void WireframePass::Render(Diligent::IDeviceContext& context,
+                           const WireframeRendererRenderState& state)
+{
+    if (state.wireframeData.empty())
+    {
+        return;
+    }
+
+    context.SetPipelineState(m_pso);
+    for (const auto& [data, mesh] : state.wireframeData)
+    {
+        m_buffer->Update(context, data);
+        const auto attribs = Diligent::DrawIndexedAttribs{
```

Comment:
Decided not to instance these because the common case right now is editor selections, which is just 0/1/2 instances.